File: backend/src/voice_agent/services/banking.py
"""Banking service — mock banking actions for the four tool functions."""

import json
from pathlib import Path
from typing import Any

from voice_agent.lib.errors import ToolExecutionError


DATA_PATH = Path(__file__).parent.parent.parent / "data" / "mock_banking.json"

_CACHED_DATA: dict[str, Any] | None = None


def _load_data() -> dict[str, Any]:
    """Load and cache mock banking data."""
    global _CACHED_DATA
    if _CACHED_DATA is None:
        with open(DATA_PATH, "r", encoding="utf-8") as f:
            _CACHED_DATA = json.load(f)
    return _CACHED_DATA


def get_customer(customer_id: str = "cust-001") -> dict[str, Any]:
    """Get customer data."""
    data = _load_data()
    customer = data["customers"].get(customer_id)
    if not customer:
        raise ToolExecutionError(f"Customer not found: {customer_id}", details={"customer_id": customer_id})
    return customer


def get_account_balance(account_id: str | None = None) -> dict[str, Any]:
    """Return balances for all accounts, or a specific account."""
    customer = get_customer()
    if account_id:
        acc = next((a for a in customer["accounts"] if a["account_id"] == account_id), None)
        if not acc:
            raise ToolExecutionError(f"Account not found: {account_id}", details={"account_id": account_id})
        return {"account": acc, "customer_name": customer["name"]}
    return {
        "accounts": customer["accounts"],
        "customer_name": customer["name"],
        "customer_name_ar": customer["name_ar"],
    }


def get_recent_transactions(limit: int = 5, account_id: str | None = None) -> dict[str, Any]:
    """Return recent transactions, optionally filtered by account."""
    customer = get_customer()
    txns = customer["transactions"]
    if account_id:
        txns = [t for t in txns if t["account_id"] == account_id]
    return {
        "transactions": txns[:limit],
        "customer_id": customer["customer_id"],
    }


def get_card_status(card_id: str | None = None) -> dict[str, Any]:
    """Return card status (default card if card_id omitted)."""
    customer = get_customer()
    if card_id:
        card = next((c for c in customer["cards"] if c["card_id"] == card_id), None)
        if not card:
            raise ToolExecutionError(f"Card not found: {card_id}", details={"card_id": card_id})
    else:
        card = customer["cards"][0] if customer["cards"] else None

    if not card:
        raise ToolExecutionError("No cards found for customer", details={"customer_id": customer["customer_id"]})

    status_labels = {
        "active": "مفعلة" if True else "active",  # Arabic labels below
        "blocked": "محظورة",
        "expired": "منتهية الصلاحية",
        "pending_activation": "في انتظار التفعيل",
    }

    # We need to determine language from context — but this is pure domain, so return both
    return {
        "card": card,
        "status_label_ar": status_labels.get(card["status"], card["status"]),
        "customer_name": customer["name"],
    }


def _update_card_status(card_id: str | None, action: str, reason: str | None = None) -> dict[str, Any]:
    """Internal: update card status. Returns updated card."""
    # For the mock, we track status changes in memory only
    customer = get_customer()
    if card_id:
        card = next((c for c in customer["cards"] if c["card_id"] == card_id), None)
    else:
        card = customer["cards"][0] if customer["cards"] else None

    if not card:
        raise ToolExecutionError("No card found", details={"card_id": card_id or "default"})

    if action == "block":
        card["status"] = "blocked"
    elif action == "unblock":
        card["status"] = "active"
    elif action == "report_lost":
        card["status"] = "blocked"

    return {
        "card": card,
        "action": action,
        "reason": reason,
        "message_ar": {
            "block": "تم حظر البطاقة بنجاح.",
            "unblock": "تم تفعيل البطاقة بنجاح.",
            "report_lost": "تم الإبلاغ عن فقدان البطاقة. سيتم إرسال بطاقة جديدة.",
        }.get(action, "تم تحديث حالة البطاقة."),
    }


def update_card_status(card_id: str | None = None, action: str = "block", reason: str | None = None) -> dict[str, Any]:
    """Tool: update card status (block/unblock/report_lost)."""
    return _update_card_status(card_id, action, reason)


def file_dispute(transaction_id: str, reason: str, contact_method: str) -> dict[str, Any]:
    """File a dispute. Appends to customer's disputes list."""
    import uuid
    from datetime import datetime, timezone

    customer = get_customer()

    # Validate transaction exists
    valid_ids = {t["id"] for t in customer["transactions"]}
    if transaction_id not in valid_ids:
        raise ToolExecutionError(
            "Transaction not found",
            details={"transaction_id": transaction_id, "valid_ids": list(valid_ids)},
        )

    dispute = {
        "id": f"disp-{uuid.uuid4().hex[:8]}",
        "transaction_id": transaction_id,
        "reason": reason,
        "contact_method": contact_method,
        "customer_id": customer["customer_id"],
        "filed_at": datetime.now(timezone.utc).isoformat(),
        "status": "pending_review",
    }

    customer["disputes"].append(dispute)

    # Persist back to JSON (simple file write for mock)
    _persist_data(_CACHED_DATA)

    reason_labels = {
        "unauthorized": "غير مصرح بها",
        "duplicate": "مكررة",
        "amount_incorrect": "مبلغ غير صحيح",
        "not_received": "لم استلم المنتج",
    }

    contact_labels = {
        "phone": "اتصال هاتفي",
        "email": "بريد إلكتروني",
        "sms": "رسالة نصية",
    }

    return {
        "dispute": dispute,
        "message": f"تم تقديم الشكوى بنجاح. رقم الشكوى: {dispute['id']}",
        "message_ar": f"تم تقديم الشكوى بنجاح. رقم الشكوى: {dispute['id']}. سنتواصل معك عبر {contact_labels.get(contact_method, contact_method)}.",
    }


def _persist_data(data: dict[str, Any]) -> None:
    """Write data back to the JSON file (mock persistence)."""
    with open(DATA_PATH, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)


# Tool handlers map — maps function name to (function, needs_args_validation)
TOOL_HANDLERS = {
    "get_account_balance": get_account_balance,
    "get_recent_transactions": get_recent_transactions,
    "update_card_status": update_card_status,
    "file_dispute": file_dispute,
}


def execute_tool(function_name: str, arguments: dict[str, Any]) -> dict[str, Any]:
    """Execute a banking tool by name with arguments dict."""
    handler = TOOL_HANDLERS.get(function_name)
    if not handler:
        raise ToolExecutionError(f"Unknown tool: {function_name}", details={"function_name": function_name})

    # Filter to only known args for each handler
    result = handler(**arguments)
    return result

File: backend/src/services/banking.ts
import { readFileSync } from 'fs'
import { join, dirname } from 'path'
import { fileURLToPath } from 'url'
import { ToolExecutionError } from '../lib/errors.js'

const __dirname = dirname(fileURLToPath(import.meta.url))
const DATA_PATH = join(__dirname, '../../data/mock_banking.json')

interface Account {
  account_id: string
  account_type: string
  balance: number
  currency: string
  status: string
}

interface Card {
  card_id: string
  last_four: string
  status: string
  expiry: string
  card_type: string
}

interface Transaction {
  id: string
  account_id: string
  amount: number
  currency: string
  description: string
  date: string
  category: string
}

interface Customer {
  customer_id: string
  name: string
  name_ar: string
  accounts: Account[]
  cards: Card[]
  transactions: Transaction[]
  disputes: Record<string, unknown>[]
}

interface MockBankingData {
  customers: Record<string, Customer>
}

let cachedData: MockBankingData | null = null

function loadData(): MockBankingData {
  if (cachedData) return cachedData
  const raw = readFileSync(DATA_PATH, 'utf-8')
  cachedData = JSON.parse(raw) as MockBankingData
  return cachedData
}

function getCustomer(customerId: string = 'cust-001'): Customer {
  const data = loadData()
  const customer = data.customers[customerId]
  if (!customer) throw new ToolExecutionError(`Customer not found: ${customerId}`, { customerId })
  return customer
}

export function getAccountBalance(accountId?: string): Record<string, unknown> {
  const customer = getCustomer()
  if (accountId) {
    const acc = customer.accounts.find((a) => a.account_id === accountId)
    if (!acc) throw new ToolExecutionError(`Account not found: ${accountId}`, { accountId })
    return { account: acc, customer_name: customer.name }
  }
  return {
    accounts: customer.accounts,
    customer_name: customer.name,
    customer_name_ar: customer.name_ar,
  }
}

export function getRecentTransactions(limit: number = 5, accountId?: string): Record<string, unknown> {
  const customer = getCustomer()
  let txns = customer.transactions
  if (accountId) txns = txns.filter((t) => t.account_id === accountId)
  return { transactions: txns.slice(0, limit), customer_id: customer.customer_id }
}

const STATUS_LABELS_AR: Record<string, string> = {
  active: 'مفعلة',
  blocked: 'محظورة',
  expired: 'منتهية الصلاحية',
  pending_activation: 'في انتظار التفعيل',
}

export function getCardStatus(cardId?: string): Record<string, unknown> {
  const customer = getCustomer()
  let card: Card | undefined
  if (cardId) {
    card = customer.cards.find((c) => c.card_id === cardId)
  } else {
    card = customer.cards[0]
  }
  if (!card) throw new ToolExecutionError('No cards found for customer', { customer_id: customer.customer_id })
  return {
    card,
    status_label_ar: STATUS_LABELS_AR[card.status] ?? card.status,
    customer_name: customer.name,
  }
}

export function updateCardStatus(
  cardId?: string,
  action: 'block' | 'unblock' | 'report_lost' = 'block',
  reason?: string,
): Record<string, unknown> {
  const customer = getCustomer()
  let card: Card | undefined
  if (cardId) {
    card = customer.cards.find((c) => c.card_id === cardId)
  } else {
    card = customer.cards[0]
  }
  if (!card) throw new ToolExecutionError('No card found', { cardId: cardId ?? 'default' })

  if (action === 'block') card.status = 'blocked'
  else if (action === 'unblock') card.status = 'active'
  else if (action === 'report_lost') card.status = 'blocked'

  const messagesAr: Record<string, string> = {
    block: 'تم حظر البطاقة بنجاح.',
    unblock: 'تم تفعيل البطاقة بنجاح.',
    report_lost: 'تم الإبلاغ عن فقدان البطاقة. سيتم إرسال بطاقة جديدة.',
  }

  return { card, action, reason, message_ar: messagesAr[action] ?? 'تم تحديث حالة البطاقة.' }
}

export function fileDispute(
  transactionId: string,
  reason: string,
  contactMethod: string,
): Record<string, unknown> {
  const customer = getCustomer()
  const validIds = new Set(customer.transactions.map((t) => t.id))
  if (!validIds.has(transactionId)) {
    throw new ToolExecutionError('Transaction not found', { transactionId, validIds: [...validIds] })
  }

  const disputeId = `disp-${Math.random().toString(36).slice(2, 10)}`
  const dispute = {
    id: disputeId,
    transaction_id: transactionId,
    reason,
    contact_method: contactMethod,
    customer_id: customer.customer_id,
    filed_at: new Date().toISOString(),
    status: 'pending_review',
  }

  customer.disputes.push(dispute)

  const contactLabels: Record<string, string> = {
    phone: 'اتصال هاتفي',
    email: 'بريد إلكتروني',
    sms: 'رسالة نصية',
  }

  return {
    dispute,
    message: `تم تقديم الشكوى بنجاح. رقم الشكوى: ${disputeId}`,
    message_ar: `تم تقديم الشكوى بنجاح. رقم الشكوى: ${disputeId}. سنتواصل معك عبر ${contactLabels[contactMethod] ?? contactMethod}.`,
  }
}

type ToolArgs = Record<string, unknown>

export function executeTool(functionName: string, args: ToolArgs): Record<string, unknown> {
  switch (functionName) {
    case 'get_account_balance':
      return getAccountBalance(args.account_id as string | undefined)
    case 'get_recent_transactions':
      return getRecentTransactions(
        (args.limit as number) ?? 5,
        args.account_id as string | undefined,
      )
    case 'update_card_status':
      return updateCardStatus(
        args.card_id as string | undefined,
        args.action as 'block' | 'unblock' | 'report_lost',
        args.reason as string | undefined,
      )
    case 'file_dispute':
      return fileDispute(
        args.transaction_id as string,
        args.reason as string,
        args.contact_method as string,
      )
    default:
      throw new ToolExecutionError(`Unknown tool: ${functionName}`, { functionName })
  }
}
